utils/docker/Makefile

```diff
@@ -76,32 +94,48 @@ DOCKER_BUILD_OPTS ?=
 ifdef FORCE_REBUILD
 	DOCKER_BUILD_OPTS += --no-cache
 endif
+DOCKER_BUILD_OPTS += --build-arg FPC_VERSION=$(FPC_VERSION)
 
 
 base:
 	$(DOCKER) build $(DOCKER_BUILD_OPTS) -t $(FPC_DOCKER_BASE_NAME) base
+	$(DOCKER) tag $(FPC_DOCKER_BASE_NAME) $(FPC_DOCKER_BASE_NAME):${FPC_VERSION}
 
 ccenv: base
 	$(DOCKER) build $(DOCKER_BUILD_OPTS) -t $(FPC_DOCKER_CCENV_NAME) ccenv
+	$(DOCKER) tag $(FPC_DOCKER_CCENV_NAME) $(FPC_DOCKER_CCENV_NAME):${FPC_VERSION}
 
 # Note: for overall consistency reasons we want the FPC code from the current repo
 # as this will rebuild each time and take a while, we don't add a dependency
-# to the build target above ...
+# to the build target above for peer and dev ...
+# To enable docker to maximize build sharing between dev and peer
+# we add DOCKER_DEV_BUILD_OPTS (e.g., to get additional packages) also to peer
 peer: base
 	(cd ${TOP}; \
          $(DOCKER) build $(DOCKER_BUILD_OPTS) -t ${FPC_DOCKER_PEER_NAME}\
+         $(DOCKER_DEV_BUILD_OPTS)\
          -f utils/docker/peer/Dockerfile\
-         --build-arg FPC_REPO_URL=file:///tmp/build-src/.git\
-         --build-arg FPC_REPO_BRANCH=$$(git rev-parse --abbrev-ref HEAD)\
+         --build-arg FPC_REPO_URL=file:///tmp/cloned-local-fpc-git-repo\
+         --build-arg FPC_REPO_BRANCH_TAG_OR_COMMIT=$$(git rev-parse HEAD)\
          --build-arg SGX_MODE=${SGX_MODE}\
          . )
+	$(DOCKER) tag $(FPC_DOCKER_PEER_NAME) $(FPC_DOCKER_PEER_NAME):${FPC_VERSION}
 
 dev: base
-	$(DOCKER) build $(DOCKER_BUILD_OPTS) -t $(FPC_DOCKER_DEV_NAME) $(DOCKER_DEV_BUILD_OPTS) dev
+	(cd ${TOP}; \
+         $(DOCKER) build $(DOCKER_BUILD_OPTS) -t $(FPC_DOCKER_DEV_NAME)\
+         $(DOCKER_DEV_BUILD_OPTS)\
+         -f utils/docker/dev/Dockerfile\
+         --build-arg FPC_REPO_URL=file:///tmp/cloned-local-fpc-git-repo\
+         --build-arg FPC_REPO_BRANCH_TAG_OR_COMMIT=$$(git rev-parse HEAD)\
+         --build-arg SGX_MODE=${SGX_MODE}\
```

Comment:
I didn't change anything when dev image docker build is called (although tried to minimize what is actually build rather than re-used from cache).  What fails in your case though shouldn't be the build but the run.  I actually did add at some time could which should have removed  a container with name fpc-development-${FPC_VERSION} iff it is dormant (you clearly don't want to kill a running one ...) but somehow it got lost; will (re-)add ...

Reply:
fyi: corresponding fixup is uploaded